backend/routes/forms/submit.py

```diff
@@ -213,3 +226,30 @@ async def send_submission_webhook(
         async with httpx.AsyncClient() as client:
             r = await client.post(form.webhook.url, json=hook)
             r.raise_for_status()
+
+    @staticmethod
+    async def assign_role(form: Form, request_user: User) -> None:
+        """Assigns Discord role to user when user submitted response."""
+        if not form.discord_role:
+            raise ValueError("Got empty Discord role ID.")
+
+        url = (
+            f"{constants.DISCORD_API_BASE_URL}/guilds/{constants.DISCORD_GUILD}"
+            f"/members/{request_user.payload['id']}/roles/{form.discord_role}"
+        )
+
+        async with httpx.AsyncClient() as client:
+            resp = await client.put(url, headers=DISCORD_HEADERS)
+            if resp.status_code == 429:  # We are rate limited
+                status = resp.status_code
+                retry_after = float(resp.headers["X-Ratelimit-Reset-After"])
+                while status == 429:
+                    await asyncio.sleep(retry_after)
+                    r = await client.put(url, headers=DISCORD_HEADERS)
+                    status = r.status_code
+                    if status == 429:
+                        retry_after = float(r.headers["X-Ratelimit-Reset-After"])
+                    else:
+                        r.raise_for_status()
+            else:  # For any other unexpected status, raise error.
+                resp.raise_for_status()
```

Comment:
This can be simplified significantly by doing:
```suggestion
            # Handle Rate Limits
            while resp.status_code == 429:
                retry_after = float(resp.headers["X-Ratelimit-Reset-After"])
                await asyncio.sleep(retry_after)
                resp = await client.put(url, headers=DISCORD_HEADERS)

            resp.raise_for_status()
```